virtual_rainforest/models/abiotic/radiation.py

```diff
@@ -0,0 +1,317 @@
+
+
+def calculate_topofcanopy_radiation(
+    shortwave_in: NDArray[np.float32],
+    elevation: NDArray[np.float32],
+    sunshine_fraction: NDArray[np.float32] = np.array(1.0, dtype=np.float32),
+    albedo_shortwave: NDArray[np.float32] = np.array(0.17, dtype=np.float32),
+) -> NDArray[np.float32]:
+    """Calculate top of canopy shortwave radiation, [J m-2].
+
+    Args:
+        shortwave_in: downward shortwave radiation, [J m-2]
+        elevation: elevation above sea level, [m]
+        sunshine_fraction: fraction of sunshine hours, between 0 (100% cloud cover)
+            and 1 (cloud free sky), default = 1
+        albedo_shortwave: shortwave albedo, default = 0.17
+
+    Returns:
+        top of canopy radiation shortwave radiation, [J m-2]
+    """
+
+    # check if sunshine fraction is in data object
+    if sunshine_fraction.size > 1:
+        sunshine_fraction = data["sunshine_fraction"]
+    else:
+        sunshine_fraction = np.ones(len(elevation), dtype=np.float32)
+        LOGGER.info("The sunshine fraction is set to default = 1")
+
+    tau = calculate_atmospheric_transmissivity(elevation, sunshine_fraction)
+    return (1.0 - albedo_shortwave) * tau * shortwave_in
+
+
+def calculate_longwave_radiation(
+    canopy_temperature: NDArray[np.float32],
+    surface_temperature: NDArray[np.float32],
+    canopy_emissivity: float = CANOPY_EMISSIVITY,
+    soil_emissivity: float = SOIL_EMISSIVITY,
+    STEFAN_BOLTZMANN_CONSTANT: float = STEFAN_BOLTZMANN_CONSTANT,
+    CELSIUS_TO_KELVIN: float = CELSIUS_TO_KELVIN,
+) -> NDArray[np.float32]:
+    """Calculate longwave emission from canopy and forest floor, [J m-2].
+
+    Args:
+        canopy_temperature: canopy temperature of n layers, [C]; the array size is
+            set to max number of layers (n_max) and filled with NaN where n < n_max
+        surface_temperature: surface soil temperature, [C]
+        canopy_emissivity: canopy emissivity, default set in config
+        soil_emissivity: soil emissivity, default set in config
+        STEFAN_BOLTZMANN_CONSTANT: Stefan-Boltzmann constant [W m-2 K-4]
+        CELSIUS_TO_KELVIN: factor to convert temperature in Celsius to absolute
+            temperature in Kelvin
+
+    Returns:
+        longwave radiation from n individual canopy layers and soil, [J m-2]
+    """
+    # longwave emission canopy
+    longwave_canopy = (
+        canopy_emissivity
+        * STEFAN_BOLTZMANN_CONSTANT
+        * (CELSIUS_TO_KELVIN + canopy_temperature) ** 4
+    )
+
+    # longwave emission surface
+    longwave_soil = (
+        soil_emissivity
+        * STEFAN_BOLTZMANN_CONSTANT
+        * (CELSIUS_TO_KELVIN + surface_temperature) ** 4
+    )
+
+    # return array of longwave radiation for all canopy layers and surface
+    return np.append(
+        longwave_canopy.transpose(),
+        longwave_soil.reshape([1, len(surface_temperature)]),
+        axis=0,
+    ).transpose()
```

Comment:
What do these operations do to the shape of the returned data? Do you just want to concatenate some vectors together?

Reply:
I tested this outside pytest and it turns our that it is not possible to concatenate arrays just as they are but the shape needs to be provided, that was also concluded by others. I'm sure there is a better way to ensure that the dimensions match but in the end, thia will also depend on the shape of the data object entries, so it is a bit hard to set this up properly...

Reply:
outdated